File: adventofcode/2024/day01/day01.rb
def part1 ls, rs
  ls.each.with_index.reduce(0) {|r, (l, i)| r + (l - rs[i]).abs}
end

def part2 ls, rs
  ls.reduce(0) {|r, l| r + l * rs.count(l)}
end

nums = readlines.map(&:split)
ls, rs, *_ = nums.transpose.map {|x| x.map(&:to_i)}
p part1 ls.sort, rs.sort
p part2 ls, rs

File: adventofcode/2024/day01/day01.hs
import Data.List (sort, transpose)
import Debug.Trace (trace)

vsorteddiffs :: ([Int], [Int]) -> Int -> Int
vsorteddiffs ([], []) r = r
vsorteddiffs ((l:ls), (r:rs)) res = vsorteddiffs (ls, rs) nres
  where nres = res + abs (l - r)

part1 :: [[Int]] -> Int
part1 ns = vsorteddiffs (sort ls, sort rs) 0
  where [ls, rs] = transpose ns

count :: Int -> [Int] -> Int
count n = length . filter (== n)

part2 :: [[Int]] -> Int
part2 ns = foldl (\r l -> r + (l * (count l rs))) 0 ls
  where [ls, rs] = transpose ns

main :: IO ()
main = interact $ unlines . map show . sequence [part1, part2] . map (map read . words) . lines

